File: src/analyzer/analyzer.py
#!/usr/bin/env python3
import subprocess
import os
import json
import logging
import multiprocessing as mp

from ..lib import utils

logger = logging.getLogger("Analyzer")

this_path = os.path.dirname(os.path.realpath(__file__))

analyzer = "target/release/analyzer"

def setup_parser(subparsers):
    parser = subparsers.add_parser("analyze", help="analyze .bc files to generate symbolic traces and features")
    parser.add_argument('--print-options', action='store_true')
    parser.add_argument('--print-call-graph', action='store_true')
    parser.add_argument('--serial', '-s', action='store_true', help='Scheduling internal jobs in serial')
    parser.add_argument('--serial-bc', action='store_true', help='Analyze on bc files one by one')
    parser.add_argument('--not-random', action='store_true', help="Do not randomly schedule the execution work")
    parser.add_argument('--slice-depth', type=int, default=0, help='Slice depth')
    parser.add_argument('--max-timeout', '-timeout', type=int, default=5, help="max timeout(s) for one work")
    parser.add_argument('--use-batch', action='store_true', help="Do use batch")
    parser.add_argument('--batch-size', type=int)
    parser.add_argument('--trace-only', action='store_true', help='Only generate symolic traces')
    parser.add_argument('--feature-only', action='store_true', help='Directly generate feature')
    parser.add_argument('--regex', action='store_true')
    parser.add_argument('--target-fn', '-target', default=[], type=str, nargs='+',
                         help='Target function(s) to be analyzed')
    parser.add_argument('--exclude-fn', '-exclude', default=[], type=str, nargs='+',
                         help='Function(s) to be ignored')
    parser.add_argument('--max-trace-per-slice', type=int, default=50)
    parser.add_argument('--max-explored-trace-per-slice', type=int, default=1000)
    parser.add_argument('--max-node-per-trace', type=int, default=2000)
    parser.add_argument('--step-in-anytime', action='store_true',
                         help='Step into the calls even if the slice depth is zero')
    parser.add_argument('--rough-mode', action='store_true', help='Save traces without satisfying path constraints')
    parser.add_argument('--bc', "-bc", type=str, default="", help='The .bc file to analyze')
    parser.add_argument('--bcdir', '-bcdir', type=str, default=None,
                         help='The directory where to save/load .bc files, default is codebase/bc-files')
    parser.add_argument('--outdir','-outdir', type=str, default=None,
                         help='The directory where to save/load output, default is codebase/cad-output')

def get_analyzer_args(bc_file, args, tmp_folder=None, extract_features=False):
    bc_name = os.path.basename(bc_file)
    base_args = [bc_file, args.outdir, '--subfolder', bc_name, '--slice-depth', str(args.slice_depth)]

    if args.print_call_graph:
        base_args += ['--print-call-graph']

    if args.print_options:
        base_args += ['--print-options']

    if args.serial:
        base_args += ['--use-serial']

    if args.not_random:
        base_args += ['--not-random-scheduling']

    if args.max_timeout:
        base_args += ['--max-timeout', str(args.max_timeout)]

    if args.use_batch:
        base_args += ['--use-batch']

    if args.batch_size:
        base_args += ['--batch-size', str(args.batch_size)]

    if args.feature_only:
        base_args += ['--feature-only']

    if args.regex:
        base_args += ['--use-regex-filter']

    if args.target_fn:
        base_args += ['--target-inclusion-filter'] + args.target_fn

    if args.exclude_fn:
        base_args += ['--target-exclusion-filter'] + args.exclude_fn

    if args.max_trace_per_slice != None:
        base_args += ['--max-trace-per-slice', str(args.max_trace_per_slice)]

    if args.max_explored_trace_per_slice != None:
        base_args += ['--max-explored-trace-per-slice', str(args.max_explored_trace_per_slice)]

    if args.max_node_per_trace != None:
        base_args += ['--max-node-per-trace', str(args.max_node_per_trace)]

    if extract_features == False:
        base_args += ['--no-feature']

    if args.step_in_anytime:
        base_args += ['--step-in-anytime']

    if args.rough_mode:
        base_args += ['--rough-mode']

    if tmp_folder != None:
        base_args += ['--metadata-file', f'{tmp_folder}/{bc_name}_metadata.json']
        base_args += ['--target-num-slices-map-file', f'{tmp_folder}/{bc_name}.json']

    return base_args

def run_analyzer_on_bc_file(bc_file, args, tmp_folder=None, only_one=False):
    logger.info(utils.color_str(f"=== Running analyzer on {bc_file}... ===", "green"))
    if only_one:
        analyzer_args = get_analyzer_args(bc_file, args, extract_features=True)
    else:
        analyzer_args = get_analyzer_args(bc_file, args, tmp_folder=tmp_folder)

    cmd = [analyzer] + analyzer_args
    ret = subprocess.run(cmd, cwd=this_path)
    if ret.returncode != 0:
        # This should rarely happen.
        # In fact, there may be other more complex reasons if it happens.
        alert_text = utils.color_str(f"Failure during handling {os.path.basename(bc_file)}.\n" \
                        f"error code: {ret.returncode}.\n" \
                        "Take command perror/errno to see the description or check it in <errno.h>.\n" \
                        "Note: The most common reason is killed by OS because of the limited resource.\n" \
                        "\tUsing -s can relieve some pressure.")
        logger.info(alert_text)
        return ""

    if only_one:
        return
    bc_name = os.path.basename(bc_file)
    output_file = args.outdir + f'/{tmp_folder}/{bc_name}.json'
    # Format: {"func_name": (has_return_type, slices_num)}
    with open(output_file) as f:
        return json.load(f)

def par_job(inputs):
    return run_analyzer_on_bc_file(inputs[0], inputs[1], inputs[2])

def combine_metadata(outdir, bc_files, tmp_folder):
    combined_metadata = {}
    for bc_file in bc_files:
        bc_name = os.path.basename(bc_file)
        output_file = outdir + f'/{tmp_folder}/{bc_name}_metadata.json'
        new_metadata = {}
        try:
            with open(output_file) as f:
                new_metadata = json.load(f)
            if combined_metadata == {}:
                combined_metadata = new_metadata
            else:
                for key in combined_metadata:
                    combined_metadata[key] += new_metadata[key]
        except:
            continue
    return combined_metadata

def generate_func_num_slices_map(output_dir, package_occurs_map, tmp_folder):
    functions = {}
    for bc_path, occurrences in package_occurs_map.items():
        bc_name = os.path.basename(bc_path)
        for func_name, (has_return_type, num_slices) in occurrences.items():
            if num_slices > 0:
                if not func_name in functions:
                    functions[func_name] = {
                        "name": func_name,
                        "package_num_slices": [],
                        "has_return_type": has_return_type
                    }
                functions[func_name]["package_num_slices"].append([bc_name, num_slices])
    ''' 
    func_num_slices_map: {
        "functions": [{"name": "function_name", 
            "has_return_type": true/false,
            "package_num_slices": [["bcfile_name_1", num_slices_1], ...]}, ...]
    }
    '''
    func_num_slices_map = {"functions": list(functions.values())}

    filename = output_dir + "/" + tmp_folder + "/ALL.json"
    with open(filename, "w") as f:
        json.dump(func_num_slices_map, f)

    return filename

def run_feature_extractor(func_num_slices_map, args):
    logger.info(utils.color_str("=== Running feature extraction... ===", "green"))
    extractor = "target/release/feature-extract"
    extractor_args = [func_num_slices_map, args.outdir]
    cmd = [extractor] + extractor_args
    ret = subprocess.run(cmd, cwd=this_path, stderr=subprocess.STDOUT)
    if ret.returncode != 0:
        logger.warn(utils.color_str("Failure during run_feature_extractor"))

def main(args):
    # Acquire the .bc files to analyze
    bc_files_to_run = utils.bc_files_to_run(args)
    log_path = os.path.join(args.outdir, "analyze_log.txt")
    utils.config_log_file(logger, log_path, "a")

    logger.info(utils.color_str("=== Starting to analyze ===", "green"))
    if len(bc_files_to_run) > 0:
        if len(bc_files_to_run) > 1 or args.trace_only:
            # Generate temporary folder that is shared among all runs
            tmp_folder_name = utils.generate_tmp_folder(args.outdir)
            if tmp_folder_name == "":
                logger.warn(utils.color_str(f"Failure during making temp directory"))
            # {"bc_file_path": {"func_name": (has_return_type, slices_num)}}
            package_occurs_map = {}

            # Run analyzer on each .bc file
            if args.serial_bc:
                for bc_file in bc_files_to_run:
                    try:
                        occurrences = run_analyzer_on_bc_file(bc_file, args, tmp_folder_name)
                        if len(occurrences) > 0:
                            package_occurs_map[bc_file] = occurrences
                    except Exception as e:
                        logger.info(e)
                        exit()
            else:
                import signal
                ignore_sigint = signal.signal(signal.SIGINT, signal.SIG_IGN)
                pool = mp.Pool(processes=mp.cpu_count(), )
                par_job_inputs = [(bc_file, args, tmp_folder_name) for bc_file in bc_files_to_run]
                try:
                    signal.signal(signal.SIGINT, ignore_sigint) # re-enable to catch sigint
                    results = pool.map_async(par_job, par_job_inputs)
                    pool.close()
                    pool.join()
                    results = results.get()
                    for i, bc_file in enumerate(bc_files_to_run):
                        if len(results[i]) > 0:
                            package_occurs_map[bc_file] = results[i]
                except KeyboardInterrupt:
                    logger.warn(utils.color_str(f"=== KeyboardInterrupt: Self-Terminated ===\n"))
                    # Kill itself and its subprocesses
                    os.killpg(os.getpid(), signal.SIGKILL)

            # Make statistics and output the total metadata in the original order
            metadata_ordered_item = ["proper_trace_count", "path_unsat_trace_count", "branch_explored_trace_count",
                "duplicate_trace_count", "no_target_trace_count", "exceeding_length_trace_count",
                "timeout_trace_count", "unreachable_trace_count", "explored_trace_count"]
            total_metadata = combine_metadata(args.outdir, bc_files_to_run, tmp_folder_name)
            ordered_metadata = dict()
            for item in metadata_ordered_item:
                ordered_metadata[item] = total_metadata[item]
            logger.info(utils.color_str("Total Metadata: " + str(ordered_metadata), "cyan"))

            if not args.trace_only:
                # Run feature extractor on all data generated
                func_num_slices_map = generate_func_num_slices_map(args.outdir, package_occurs_map, tmp_folder_name)
                run_feature_extractor(func_num_slices_map, args)

            # Remove the tmp folder
            utils.rmdir(args.outdir + "/" + tmp_folder_name)

        else: # There is only one bc file to be analyzed
            bc_file = bc_files_to_run[0]
            run_analyzer_on_bc_file(bc_file, args, only_one=True)
        logger.info(utils.color_str("=== Finished ===\n", "green"))
    else:
        logger.warn(utils.color_str(f"no .bc files in {args.bcdir}\n"))

File: src/analyzer/src/bin/analyzer.rs
use llir::Module;
use std::collections::HashMap;
use std::path::PathBuf;
use structopt::StructOpt;

use analyzer::{call_graph::*, feature_extraction::*, options::*, slicer::*, symbolic_execution::*, utils::*};

#[derive(StructOpt, Debug, Clone)]
#[structopt(name = "analyzer")]
pub struct Options {
    //************************** BasicOption & GeneralOptions & IOOption **************************//
    #[structopt(index = 1, required = true, value_name = "INPUT")]
    pub input: String,

    #[structopt(index = 2, required = true, value_name = "OUTPUT")]
    pub output: String,

    #[structopt(long, takes_value = true, value_name = "SUBFOLDER")]
    pub subfolder: Option<String>,

    // Print call graph
    #[structopt(long)]
    pub print_call_graph: bool,

    #[structopt(long)]
    pub print_options: bool,

    // Serialize execution rather than parallel
    #[structopt(short = "s", long)]
    pub use_serial: bool,

    // Use batch execution. Especially useful when applying to large dataset
    #[structopt(long)]
    pub use_batch: bool,

    // The number of slices inside each batch
    #[structopt(long, takes_value = true, default_value = "50", value_name = "BATCH_SIZE")]
    pub batch_size: usize,

    // The file path for dumping the metadata of generated traces
    #[structopt(long, takes_value = true, value_name = "METADATA_FILE")]
    pub metadata_file: Option<String>,

    // The file path for dumping target-num-slices-map
    #[structopt(long, takes_value = true, value_name = "TARGET_NUM_SLICES_MAP")]
    pub target_num_slices_map_file: Option<String>,

    #[structopt(long)]
    pub no_feature: bool,

    #[structopt(long)]
    pub feature_only: bool,
    //************************** BasicOption & GeneralOptions & IOOption **************************//

    //***************************************** SliceOptions *************************************//
    #[structopt(
        short = "d",
        long,
        takes_value = true,
        default_value = "1",
        value_name = "SLICE_DEPTH"
    )]
    pub slice_depth: usize,

    // Maximum number of blocks per slice
    #[structopt(long, takes_value = true, default_value = "1000", value_name = "MAX_AVG_NUM_BLOCKS")]
    pub max_num_blocks: usize,

    // Use regex in the filters
    #[structopt(long)]
    pub use_regex_filter: bool,

    #[structopt(long, takes_value = true, value_name = "INCLUDE_TARGET")]
    pub target_inclusion_filter: Option<Vec<String>>,

    #[structopt(long, takes_value = true, value_name = "EXCLUDE_TARGET")]
    pub target_exclusion_filter: Option<Vec<String>>,
    //***************************************** SliceOptions *************************************//

    //*********************************** SymbolicExecutionOptions *******************************//
    // Maximum time per work can execute
    #[structopt(long, takes_value = true, default_value = "5", value_name = "MAX_TIMEOUT")]
    pub max_timeout: usize,

    #[structopt(long, takes_value = true, default_value = "5000", value_name = "MAX_NODE_PER_TRACE")]
    pub max_node_per_trace: usize,

    #[structopt(
        long,
        takes_value = true,
        default_value = "1000",
        value_name = "MAX_EXPLORED_TRACE_PER_SLICE"
    )]
    pub max_explored_trace_per_slice: usize,

    // The maximum number of generated trace per slice
    #[structopt(long, takes_value = true, default_value = "50", value_name = "MAX_TRACE_PER_SLICE")]
    pub max_trace_per_slice: usize,

    // Step into the calls even if the slice depth is zero
    #[structopt(long)]
    pub step_in_anytime: bool,

    // Roughly exploring without checking SAT
    #[structopt(long)]
    pub rough_mode: bool,

    // Random schedule the execution work
    #[structopt(long)]
    pub not_random_scheduling: bool,
    //*********************************** SymbolicExecutionOptions *******************************//
}

impl GeneralOptions for Options {
    fn use_serial(&self) -> bool {
        self.use_serial
    }

    fn use_batch(&self) -> bool {
        self.use_batch
    }
}

impl IOOptions for Options {
    fn input_path(&self) -> PathBuf {
        PathBuf::from(&self.input)
    }

    fn output_path(&self) -> PathBuf {
        PathBuf::from(&self.output)
    }

    fn basename_of_bc_file(&self) -> Option<&str> {
        match &self.subfolder {
            Some(subfolder) => Some(&subfolder),
            None => None,
        }
    }
}

impl Options {
    fn metadata_file_path(&self) -> Option<PathBuf> {
        if let Some(filename) = &self.metadata_file {
            Some(self.output_path().join(filename))
        } else {
            None
        }
    }

    fn target_num_slices_map_path(&self) -> Option<PathBuf> {
        if let Some(filename) = &self.target_num_slices_map_file {
            Some(self.output_path().join(filename))
        } else {
            None
        }
    }

    fn num_slices(&self, target: &str) -> usize {
        match std::fs::read_dir(self.slice_target_dir(target)) {
            Ok(dirs) => dirs.count(),
            _ => 0,
        }
    }
}

impl SlicerOptions for Options {
    fn slice_depth(&self) -> usize {
        self.slice_depth as usize
    }

    fn target_inclusion_filter(&self) -> &Option<Vec<String>> {
        &self.target_inclusion_filter
    }

    fn target_exclusion_filter(&self) -> &Option<Vec<String>> {
        &self.target_exclusion_filter
    }

    fn use_regex_filter(&self) -> bool {
        self.use_regex_filter
    }

    fn max_num_blocks(&self) -> usize {
        self.max_num_blocks
    }
}

impl SymbolicExecutionOptions for Options {
    fn slice_depth(&self) -> usize {
        self.slice_depth
    }

    fn max_timeout(&self) -> usize {
        self.max_timeout
    }

    fn max_node_per_trace(&self) -> usize {
        self.max_node_per_trace
    }

    fn max_explored_trace_per_slice(&self) -> usize {
        self.max_explored_trace_per_slice
    }

    fn max_trace_per_slice(&self) -> usize {
        self.max_trace_per_slice
    }

    fn step_in_anytime(&self) -> bool {
        self.step_in_anytime
    }

    fn is_rough(&self) -> bool {
        self.rough_mode
    }

    fn not_random_scheduling(&self) -> bool {
        self.not_random_scheduling
    }
}

impl FeatureExtractorOptions for Options {}

fn main() -> Result<(), String> {
    let options = Options::from_args();
    if options.print_options {
        println!("{:?}", options);
    }

    // Load a logging context
    let mut logging_ctx = LoggingContext::new(&options)?;

    // Load the byte code module and generate analyzer context
    logging_ctx.log_loading_bc()?;
    let llctx = llir::Context::create();
    let llmod = llctx
        .load_module(&options.input_path())
        .map_err(|err| err.to_string())?;

    // Generate call graph
    logging_ctx.log_generating_call_graph()?;
    let call_graph = CallGraph::from_module(&llmod);
    if options.print_call_graph {
        call_graph.print();
    }

    // Finding call edges
    logging_ctx.log_finding_call_edges()?;
    let target_edges_map = TargetEdgesMap::from_call_graph(&call_graph, &options)?;

    // Check if we need to do the symbolic execution
    let occurrences = if !options.feature_only {
        // Generate slices from the edges
        logging_ctx.log_generated_call_edges(target_edges_map.num_elements())?;
        let target_slices_map = TargetSlicesMap::from_target_edges_map(&target_edges_map, &call_graph, &options);
        let occurrences = keyed_num_elements(&target_slices_map, &llmod);
        // Dump slices to file
        logging_ctx.log_generated_slices(target_slices_map.num_elements())?;
        target_slices_map.dump(&options);

        // Divide target slices into batches
        let num_batches = 1 + target_slices_map.num_elements() / options.batch_size;
        let batchmap = target_slices_map.batches(options.use_batch, options.batch_size);
        logging_ctx.log_dividing_batches(options.use_batch, num_batches)?;
        let mut global_metadata = MetaData::new();
        for (i, target_slices_map) in batchmap {
            logging_ctx.log_executing_batch(
                i,
                num_batches,
                options.use_batch,
                target_slices_map.len(),
                target_slices_map.num_elements(),
            )?;
            // Symbolic execution
            let sym_exec_ctx = SymbolicExecutionContext::new(&options);
            let metadata = sym_exec_ctx.execute_target_slices_map(target_slices_map);
            global_metadata = global_metadata.combine(metadata.clone());
            logging_ctx.log_finished_execution_batch(i, options.use_batch, metadata)?;
        }
        logging_ctx.log_finished_execution(options.use_batch, &global_metadata)?;
        // Save metadata to a temporary file
        if let Some(filename) = options.metadata_file_path() {
            global_metadata.dump(filename)?;
        }

        occurrences
    } else {
        // If not, we directly load slices information from file
        load_target_num_slices_map(target_edges_map, &llmod, &options)
    };

    // Dump the occurrences(format: {"func_name": (has_return_type, slices_num), ...}) to file
    if let Some(filename) = options.target_num_slices_map_path() {
        logging_ctx.log_dump_target_num_slices_map(&filename)?;
        occurrences.dump(filename)?;
    }

    if !options.no_feature {
        // Directly extract features
        logging_ctx.log_extracting_features()?;
        let feature_extract_ctx = FeatureExtractionContext::new(&llmod, occurrences, &options)?;
        feature_extract_ctx.extract_features(&mut logging_ctx);
        logging_ctx.log_finished_extracting_features()?;
    }

    Ok(())
}

// (func_name, (has_return_type, slices_num))
fn keyed_num_elements(target_slices_map: &TargetSlicesMap, module: &Module) -> HashMap<String, (bool, usize)> {
    let func_types = module.function_types();
    let mut result = HashMap::new();
    for (target, value) in target_slices_map {
        result.insert(
            target.clone(),
            (func_types[&target.clone()].has_return_type(), value.len()),
        );
    }
    result
}

// (func_name, (has_return_type, slices_num))
fn load_target_num_slices_map(
    target_edges_map: TargetEdgesMap,
    module: &Module,
    options: &Options,
) -> HashMap<String, (bool, usize)> {
    let func_types = module.function_types();
    target_edges_map
        .into_iter()
        .map(|(target, _)| {
            let num_slices = options.num_slices(&target);
            (target.clone(), (func_types[&target].has_return_type(), num_slices))
        })
        .collect()
}
